runtime/include/chpl-mem.h

```diff
@@ -169,11 +164,13 @@ void* chpl_mem_array_alloc(size_t nmemb, size_t eltSize, c_sublocid_t subloc,
     if (p == NULL) {
       p = chpl_mem_allocMany(nmemb, eltSize, CHPL_RT_MD_ARRAY_ELEMENTS,
                              lineno, filename);
-      do_localize = !chpl_mem_localizes();
+      // TODO is this right?
```

Comment:
Would you mind leaving chpl_mem_localizes() and the associated logic?  While it's true that with this change it collapses to false in all cases including jemalloc and therefore could be removed, as soon as jemalloc can allocate already-localized memory it will be true again.  (Which is worthwhile because then we can avoid the chpl_topo_setMemLocality() below.)  And the check against the subloc value is handled later if we're going to localize at all.

Reply:
I'm not sure I'm following. What are we going to need that for? Not for dynamic registration or chunks, right?

Reply:
(personally I'd like to remove it still, it's trivial to re-add in the future if we do need it)

Reply:
It seems like it would be useful for chunks.  With the chunk work we can localize (via the topo layer) each chunk that we add to an arena so that it's in the same NUMA domain as the thread which is extending that arena (which is presumably also running in the NUMA domain).  Subsequent allocations from the same or other tasks running on that thread will then also have good affinity.  Given this, we want to say that the memory layer localizes allocations, so we can avoid calling chpl_topo_whatever() to post-localize the allocation later in chpl_mem_array_alloc().

Reply:
We can add it back later, yes.

Reply:
Ah ok. Good point. I forgot that nic registering will pin to numa 0 unless it's already pinned. For some reason I was thinking it would still use sched_affinity to pin to whatever domain the current thread is running on. 

So I definitely agree that we'll want to localize a chunk, but I'm not sure there will be any clients that need to know that information. Currently nobody calls chpl_mem_array_alloc() with anything other than `c_sublocid_all` or `c_sublocid_none`, so the chpl_topo_setMemLocality() is never actually called, right? Do we think that will change in the near future?

Reply:
(and I don't mean to make a big deal out of this, if you think it's important I'm happy to leave it)

Reply:
Ah.  Maybe not.  Maybe we can just yank out the whole do_localize-based if-stmt that controls the post-alloc topo localization calls further down, and rely on first-touch in all cases where it can have an effect in the first place.

A bit of background: chpl_mem_localizes() used to be called chpl_mem_impl_localizes(), which is arguably a better name because this is a function that the general code here maybe, or maybe not) needs to call but nobody else does.  So even if we keep it it doesn't need to be defined as part of the publicized interface but it does need to be provided by every implementation (even if vacuously, by not overriding the default), so the it's available to be used by the interface code.